src/lender/tranche/operator/proportional.sol

```diff
@@ -0,0 +1,171 @@
+// Copyright (C) 2020 Centrifuge
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.5.15 <0.6.0;
+
+import "ds-note/note.sol";
+import "tinlake-math/math.sol";
+import "tinlake-auth/auth.sol";
+
+contract TrancheLike {
+    function supply(address usr, uint currencyAmount, uint tokenAmount) public;
+    function redeem(address usr, uint currencyAmount, uint tokenAmount) public;
+    function tokenSupply() public returns (uint);
+}
+
+contract AssessorLike {
+    function calcAndUpdateTokenPrice(address tranche) public returns(uint);
+    function supplyApprove(address tranche, uint currencyAmount) public returns(bool);
+    function redeemApprove(address tranche, uint currencyAmount) public returns(bool);
+    function tokenAmountForONE() public returns(uint);
+}
+
+contract DistributorLike {
+    function balance() public;
+}
+
+contract ProportionalOperator is Math, DSNote, Auth  {
+    TrancheLike public tranche;
+    AssessorLike public assessor;
+    DistributorLike public distributor;
+
+    // lender mappings
+    // each value in a own map for gas-optimization
+    mapping (address => uint) public supplyMaximum;
+    mapping (address => uint) public tokenReceived;
+    // helper we could also calculate it
+    mapping (address => uint) public tokenRedeemed;
+
+    // expressed in totalCurrencyReturned notation
+    mapping (address => uint) public currencyRedeemed;
+
+    // expressed in totalPrincipalReturned notation
+    mapping (address => uint) public principalRedeemed;
```

Comment:
suggestion: // principal amount of which the investor's share in the pool has already been redeemed